sdks/cpp/common/examples/full_service/full_service.cpp

```diff
@@ -323,13 +319,14 @@ class CatenaServiceImpl final : public catena::CatenaService::AsyncService {
 
                 case CallStatus::kProcess:
                     new GetValue(service_, dm_, ok);
-                    if (ok) {
+                    if (ok){
                         context_.AsyncNotifyWhenDone(this);
-                        if (validateRequest_() && !context_.IsCancelled()) {
+                        if (validateRequest_()) {
                             try {
+                                std::vector<std::string> clientScopes = getScopes(context_);
```

Comment:
it's understandable why you made client scopes a vector of strings. however, it could be better / more efficient to code them as a bit-field. with monitor being bits 0 & 1, operate 2 & 3 and so on.

operate read access is signaled by bit 2 being 1, no read access signaled by it being 0, write access is granted when bit 3 is 1, and prohibited when it is zero.

don't do this now - until it's proven that hauling strings around and parsing them for every parameter access is unduly burdensome - "premature optimization" is a curse!